lib/rules/no-shadow.js

```diff
@@ -165,6 +257,10 @@ module.exports = {
                     !isOnInitializer(variable, shadowed) &&
                     !(options.hoist !== "all" && isInTdz(variable, shadowed))
                 ) {
+                    if (isInaccessible(shadowed, variable)) {
+                        continue;
+                    }
+
```

Comment:
This fix probably shouldn't change how the `"hoist": "all"` option works, since it reports shadowing even if the shadowed variable is defined clearly after the inner one:

```js
/*eslint no-shadow: ["error", { "hoist": "all" }] */

people.find(person => person.name === "John"); // 'person' is already declared in the upper scope  no-shadow

const person = "foo";
```

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tc2hhZG93OiBbXCJlcnJvclwiLCB7IFwiaG9pc3RcIjogXCJhbGxcIiB9XSAqL1xuXG5wZW9wbGUuZmluZChwZXJzb24gPT4gcGVyc29uLm5hbWUgPT09IFwiSm9oblwiKTtcblxuY29uc3QgcGVyc29uID0gXCJmb29cIjsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjExLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19)